tls: Run separate cockpit-ws instance for each client certificate

martinpitt opened the pull request:
See individual commits for details.

 - [x] Factorize units in Makefile: PR #13011  
 - [x] Fix ws socket instance leak: PR #13014 
 - [x] Fix unbounded number of instance sockets (DoS)
 - [x] Eliminate cert → socket name map in cockpit-tls to get along with instance sockets going away (caught by new `testUnitLifecycle` check)
 - [x] Add SELinux policy for new cockpit-wsinstance-factory: https://github.com/fedora-selinux/selinux-policy-contrib/pull/161

martinpitt commented:
The previous result set looked great. I adjusted testUnitLifecycle.

martinpitt commented:
@mvollmer: This PR has "Lis" written all over it, but in her absence, could you please review this? I think it's the last blocker before I can dust off the [smart card auth](https://github.com/cockpit-project/cockpit/pull/11421) for real :grin:

martinpitt commented:
This is now blocked, as we revert the epoll → thread cockpit-tls conversion in PR #12988 due to issue #12977. I could rewrite it to the epoll layout, but let's not waste work here -- I'm sure @allisonkarlitskaya can fix the bug in threading in no time.

martinpitt commented:
Now that #12977 is out of the way, this is ready for review again.

mvollmer commented:
> Now that #12977 is out of the way, this is ready for review again.

(#12988?)

Oh, so this doesn't depend on the threaded I/O at all?  Makes sense... :-)

martinpitt commented:
@allisonkarlitskaya : You mentioned some initial thoughts on IRC earlier today. Some thoughts for the ones I can remember:

 * You wanted the "no cert" instance be handled specially, without the factory. Now, I actually like having *one* mechanism to instantiate https ws, the whole structure is already complicated enough. I really veto adding yet a fourth wsinstance unit type, as it would invoke cockpit-ws in exactly the same way. Some middle ground would be that cockpit.service spawns a "static" cockpit-wsinstance-https@nocert.socket and cockpit-tls can then hardcode that unix socket name. But the one extra factory call during cockpit-tls'es life time is rather negligible IMHO, and it would add (a tiny bit of) extra code to cockpit-tls that we don't strictly need.

 * There is the possibility of DoSing the machine by sending requests with $OMGlots of different client certificates. This is mitigated by (1) this functionality isn't actually enabled yet, so if we just land this there is no  DoS yet -- there will only ever be a https@0 instance. And (2) client cert auth will *not* be enabled by default, you will need to enable it in cockpit.conf explicitly. We can document this possible case and suggest mitigations. We could even apply some reasonable boundaries by default, like setting `TriggerLimit{IntervalSec,Burst}=` in the factory or putting a CPU and Memory limit to the group of https@* instances (i. e. their parent slice).

 * You mentioned some idea about getting rid of the cert → unix_path map in cockpit-tls server.c and replace it with something else. I didn't quite get what you meant there, do you mind jotting this down? Or should we bjn about it?

mvollmer commented:
As discussed earlier, this scheme will accumulate more and more active socket units (instances of the `cockpit-wsinstance-https@.socket` template).

These socket units are `PartOf` cockpit.service, so they should go away once cockpit.service is stopped.  This is indeed the case, but only when running `systemctl stop cockpit` explicitly.  When cockpit.service goes inactive becasue cockpit-tls exits, this cleanup is not done and the sockets remain active.

To see this, open the login page of Cockpit in a browser, but don't log in.  There will be a new `cockpit-wsinstance-https@N.socket` for this.  Then let cockpit-tls time out.  The cockpit.service will be inactive, but the socket is still there.

At this point, reload the browser.  A new socket unit will appear.

Thus, the number of active socket units will go up without bounds until the machine is rebooted.

mvollmer commented:
Alternatively, maybe this counts as a bug in systemd?  Maybe it should do the PArtOf cleanup also when a service exits on its own.

mvollmer commented:
Replacing `PartOf` with `BindsTo` seems to do the trick, but I can't say I understand the nuances of systemd relationships.  It seems "complicated".

mvollmer commented on src/ws/cockpit-wsinstance-https@.socket.in:
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=Socket for Cockpit Web Service https instance %I
+PartOf=cockpit.service
```
`PartOf` doesn't seem strong enough; the sockets stay around when cockpit-tls times out.  Replacing this with `BindsTo` seems to work.

martinpitt replied:
BindsTo has the effect that e. g. starting `cockpit-wsinstance-http.socket` manually will also activate cockpit.service. This feels a bit weird, but (1) it's not really harmful, and (2) who does that. I first tried with `StopWhenUnneeded=`,  but that's too strong. So indeed I agree that BindsTo= is right. This is also an issue on current master, although of course it's strictly bounded there. I fixed this with a test in #13014.

martinpitt replied:
I updated testUnitLifecycle to spot the https@ socket leak as well, fixed that, and pushed on top of #13011 and #13014.

martinpitt commented:
Rebased to current master after landing the two prerequisites.

mvollmer commented on src/tls/connection.c:
```diff
@@ -344,30 +356,155 @@ buffer_read_from_tls (Buffer           *self,
+
+  r = snprintf (factory_sockaddr.sun_path, sizeof factory_sockaddr.sun_path,
+                "%s/https-factory.sock", parameters.wsinstance_sockdir);
+  assert (r < sizeof factory_sockaddr.sun_path);
```
Are we ok with using assert here?

mvollmer replied:
A call to warn and returning NULL might be more robust since wsinstance_sockdir is external input, no?

mvollmer replied:
I see, makes sense!  (Also, we wont ever defined NDEBUG, I guess.)

mvollmer commented:
`testUnitLifecycle (check_connection.TestConnection)` has failed on Ubuntu: https://logs-https-cockpit.apps.ci.centos.org/logs/pull-12972-20191021-082828-96096cd1-cockpit-project-cockpit--ubuntu-1804/log.html#253-2

martinpitt commented:
Thanks @mvollmer. I'll look at the test. Also, I want @allisonkarlitskaya to review this too, so please let's not land this for now.

martinpitt commented:
It seems the older systemd on Ubuntu 18.04 does not recycle the https@0 name after stopping/restarting, but goes on to @1. We don't actually care about what particular name the instances have, just how many of them are active. I updated the test to just assert the count instead of their precise name.

mvollmer commented:
I am "green" with this, but as @martinpitt says, let's have @allisonkarlitskaya approve it formally.

martinpitt commented:
@allisonkarlitskaya and I discussed this at length. The main issue with this is that the number of https@.socket grows with the number of certificates that ever connected, which is unbounded and a DoS. This functionality isn't actually exposed with just this PR, and even once we do expose it, it will need to be enabled in cockpit.conf explicitly. However, we have an idea how to fix this, so let's just do that right now.

I added two more integration tests that cover socket permissions and recreate the above DoS, and my approach for restricting the number of instances. This breaks the assumption in cockpit-tls that https@*.socket instances never go away, so adding "needswork". Lis is working on that part (eliminate the cert → socket map entirely and make the socket instance names predictable).

I used FIXUPs for easier review. @allisonkarlitskaya , feel free to add your bits on top directly in my branch.

martinpitt commented:
I'm still working on stabilizing `testHttpsInstanceDoS()`, but I think I got it now. `testUnitLifecycle` was stable now, so I extended it to catch the new cockpit-tls failure once a https instance goes down. That's the bit that we still need to fix with @allisonkarlitskaya 's new factory. This should now fail everywhere.

martinpitt commented:
@allisonkarlitskaya : I [sent a SELinux update PR](https://github.com/fedora-selinux/selinux-policy-contrib/pull/161) and added a local hack to the spec for the time being. I also rebased this to current master and squashed it; I still kept two commits, perhaps we want to squash them into one for the final version, but it works like that for me.

Note that I didn't actually review your commit in detail, just looking at test failures right now.

martinpitt commented:
I fixed the [debian/ubuntu netcat issue](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-12972-20191104-215454-66785163-cockpit-project-cockpit--debian-stable/log.html#250-2).

martinpitt commented on src/tls/wsinstance-factory.c:
```diff
@@ -0,0 +1,151 @@
+      debug (FACTORY, "sd_bus_wait(%llu)", (long long) remaining);
+      r = sd_bus_wait (bus, remaining);
+      if (r < 0)
+        errx (1, "Error while waiting for bus: %s", strerror (-r));
```
Thinking aloud: This could return with `r == 0`, i. e. it timed out without seeing a result. This should be okay though, as the following sd_bus_process would then just not do anything.

martinpitt commented on src/tls/wsinstance-factory.c:
```diff
@@ -0,0 +1,151 @@
+        errx (1, "Error while waiting for bus: %s", strerror (-r));
+
+      debug (FACTORY, "sd_bus_process():");
+      while (sd_bus_process (bus, NULL))
```
Again, let's abort on < 0 here.

martinpitt commented on src/tls/wsinstance-factory.c:
```diff
@@ -0,0 +1,151 @@
+    errx (1, "Must be spawned from a systemd service on a socket with Accept=yes %s", fdnames[0]);
+
+  if (!recv_all (SD_LISTEN_FDS_START, fingerprint, sizeof fingerprint, 10 * 1000000))
+    errx (1, "Didn't receive fingerprint");
```
Let's do `err()` here to get an errno.

martinpitt commented on src/tls/wsinstance-factory.c:
```diff
@@ -0,0 +1,151 @@
+
+  /* This is our job. */
+  debug (FACTORY, "  -> sending result.");
+  send_all (SD_LISTEN_FDS_START, result, strlen (result), 5 * 1000000);
```
Please `warn()` (or even `err()`) if this fails.

allisonkarlitskaya commented on src/tls/socket-activation-helper.c:
```diff
@@ -145,23 +146,30 @@ spawn_cockpit_ws (const char *ws_path, int fd,
-  debug (HELPER, "https factory: sending instance %s", sock_name);
+  debug (HELPER, "  -> reading instance name... ");
+  if (!recv_all (fd, fingerprint, sizeof fingerprint, 10 * 1000000))
+    errx (EXIT_FAILURE, "failed to read fingerprint");
```
`recv_all()` can return `false` after successful recv() (for example if the buffer is full).

martinpitt replied:
Ack. So maybe {send,recv}_all() should do the warn() on errors for themselves? (Although that's a bit weird as a general API, but it doesn't need to be that general)

allisonkarlitskaya commented on src/tls/socket-io.c:
You can also do only one recv_all(): it relies on receiving EOF.  It's symmetric in this regard.

martinpitt replied:
Indeed, my bad! Closing then.

allisonkarlitskaya commented on src/tls/socket-io.c:
I'm not sure I understand what kind of error vsnprintf() might return.  it doesn't do any IO.

martinpitt replied:
Sure, there are no "obvious" errors. But as the manpage documents it, let's assert r > 0?

allisonkarlitskaya commented on src/tls/utils.h:
```diff
@@ -25,11 +25,12 @@
-#define DEBUG_IOVEC 1
+#define DEBUG_POLL 0
+#define DEBUG_BUFFER 0
+#define DEBUG_IOVEC 0
```
I did.

allisonkarlitskaya commented on src/tls/test-socket-activation-helper.sh:
```diff
@@ -34,16 +34,18 @@ expect_curl() {
-    OUT=$(nc --unixsock "$SOCKET_DIR/$1" </dev/null)
+# args: <instance> <expected output>
+expect_start() {
+    OUT=$(./wsinstance-start "$1" "$SOCKET_DIR")
```
I was getting unreliable behaviour with `nc` (probably related to sometimes needing `-N` and other times not).  I had also (separately) developed this tool to use during testing.  Using it from the testcase came later...

allisonkarlitskaya commented on src/tls/wsinstance-factory.c:
```diff
@@ -0,0 +1,151 @@
+    errx (1, "Didn't receive fingerprint");
+
+  r = sd_bus_open_system (&bus);
+
```
I changed a bunch of these.

allisonkarlitskaya commented on src/tls/wsinstance-factory.c:
```diff
@@ -0,0 +1,151 @@
+
+  debug (FACTORY, "  -> path: %s, result: %s", path, result);
+
+  if (*job_path && strcmp (path, *job_path) != 0)
```
Great catch!

The problem is actually the other way around: it's possible that we're still looping after *job_path gets set to NULL.  I added this check later in response to debugging and clearly wasn't careful enough.  Thanks!

martinpitt commented on src/tls/connection.c:
```diff
@@ -429,28 +436,38 @@ static bool
-  return connect (self->ws_fd, (struct sockaddr *) &addr, sizeof addr) == 0;
+  if (connect (self->ws_fd, (struct sockaddr *) &addr, sizeof addr) != 0)
+    {
+      warn ("connect(%s) failed on the second attempt", addr.sun_path);
```
%m

allisonkarlitskaya replied:
`warn()` does it for itself

martinpitt replied:
GitHub bug report: There's no emoji for :brown_paperbag or :shame_on_me

martinpitt commented on src/tls/socket-io.c:
```diff
@@ -76,17 +104,41 @@ wait_for_io (struct timespec *start,
+ *
+ * Attempts to read up to @size - 1 bytes from the connected stream
+ * socket @fd, followed by EOF.  On success, a nul terminator is
+ * inserted after the last byte and the number of bytes read (which
```
It would make more sense if recv_all wouldn't do that, and recv_alnum() adds the nul-terminator instead.

allisonkarlitskaya replied:
I tried to implement it this way, but quickly remembered my reason for not doing this: after we recv() the normal data, we need to recv() once more to ensure that we get a 0 (ie: EOF).  We need at least one extra byte to do that into, and the byte where we would insert the nul terminator is extremely convenient for that purpose.

Otherwise I'd need to have a separate `char b` variable or something and a separate `recv()` that tries to read into that to make sure we get EOF.

This is definitely "clever".  I could add a comment about it.

martinpitt commented on src/tls/socket-io.c:
```diff
@@ -95,32 +147,90 @@ recv_all (int     fd,
             continue;
 
-          return false;
+          warn ("recv_all() failed");
```
%m so that we know why

allisonkarlitskaya replied:
```
       The err(), verr(), warn(), and vwarn() functions append an error message
       obtained  from  strerror(3) based on the global variable errno, preceded
       by another colon and space unless the fmt argument is NULL.
```

allisonkarlitskaya commented on Makefile.am:
```diff
@@ -189,7 +189,7 @@ WEBPACK_DEPS = $(WEBPACK_PACKAGES:%=dist/%/Makefile.deps)
 WEBPACK_MANIFEST_IN = $(WEBPACK_PACKAGES:%=pkg/%/manifest.json.in)
 
-noinst_SCRIPTS += $(WEBPACK_STAMPS) $(WEBPACK_INSTALL)
+noinst_SCRIPTS += $($(NO_BUILD_WEBPACK)WEBPACK_STAMPS) $($(NO_BUILD_WEBPACK)WEBPACK_INSTALL)
```
I actually wanted to propose this as a separate patch.  Can you link me to the discussion?

martinpitt replied:
That was PR #12798

allisonkarlitskaya commented on Makefile.am:
```diff
@@ -460,3 +460,6 @@ include src/tls/Makefile-tls.am
 include containers/Makefile.am
+
+disable-webpack:
+	touch -d 2050-01-01 $(WEBPACK_STAMPS)
```
This one was really a mistake :)

allisonkarlitskaya commented on src/tls/connection.c:
```diff
@@ -429,28 +436,38 @@ static bool
 
-  debug (CONNECTION, "  -> failed.  Requesting activation.");
+  if (errno != ENOENT && errno != ECONNREFUSED)
+    warn ("connect(%s) failed on the first attempt", addr.sun_path);
```
See other comments about `warn()`

allisonkarlitskaya commented on src/tls/connection.c:
```diff
@@ -358,23 +358,30 @@ request_dynamic_wsinstance (const char *fingerprint)
   if (fd == -1)
-    goto out;
+    {
+      warn ("socket() failed: %s", strerror (errno));
```
Oops.  I'm actually printing the error twice here, since `warn()` will add its own...

martinpitt commented:
I squashed the fixup, no further changes. From my POV this is :+1: now, but I can't formally approve as it's my PR.

martinpitt commented:
@mvollmer : @allisonkarlitskaya and me are now happy enough with this, but we are also now both at the point were we just can't stare at the same patches any more with full precision. Could you please have a final look? Thanks!

mvollmer commented on src/tls/connection.c:
```diff
@@ -357,182 +345,173 @@ buffer_read_from_tls (Buffer           *self,
+}
+
+static bool
+connection_connect_to_static_wsinstance (Connection *self)
```
This is only for http connections, and https-without-client-cert is also treated as a dynamic instance, right?  This had me confused for a second.  Would `connection_connect_to_http_wsinstance` be clearer?

martinpitt replied:
I thought about this as well. Then also _dynamic → _https. I have no strong opinion, if it's clearer I'm happy to get it renamed.

allisonkarlitskaya replied:
I prefer dynamic because it's what the function is actually doing.  Also, the https-without-client-cert also used to be handled under static (which is where the names come from) and i feel like this is an argument for how minor changes to our approach would continue to work with the static/dynamic names, but if we had http/https names, we'd need to change them...

mvollmer commented on src/ws/cockpit-wsinstance-https-factory@.service.in:
```diff
@@ -1,10 +1,7 @@
 [Unit]
```
Why is this still a template?

martinpitt replied:
The https-factory.socket still has `Accept=true`, which simplifies what the factory process has to do -- in particular, it can deal with one request at a time, do its thing, and die again. It can also skip the whole accept() stuff, but the biggest win is that it doesn't need to be a daemon which can serve arbitrarily many requests. This is quite important for this security sensitive bit.

mvollmer replied:
Ahh, and Accept=true needs the service to be a template.  Thanks!

mvollmer commented:
People get exposed to the full systemd unit complexity even for the simplest case of a https connection without client certificate:

```
cockpit-wsinstance-https@e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.service loaded active running   Cockpit Web Service https instance e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855           
cockpit.service                                                                                   loaded active running   Cockpit Web Service                                                                                           
cockpit-wsinstance-http-redirect.socket                                                           loaded active listening Socket for Cockpit Web Service http-redirect instance                                                         
cockpit-wsinstance-http.socket                                                                    loaded active listening Socket for Cockpit Web Service http instance                                                                  
cockpit-wsinstance-https-factory.socket                                                           loaded active listening Socket for Cockpit Web Service https instance factory                                                         
cockpit-wsinstance-https@e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.socket  loaded active running   Socket for Cockpit Web Service https instance e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
cockpit.socket                                                                                    loaded active running   Cockpit Web Service Socket                                                                                    
```

Could we keep the fingerprint out of the cert-less instance?  Could we create the http and http-redirect socket units also on demand via the factory?  Then the common case could look like

```
cockpit-wsinstance@https.socket
cockpit-wsinstance@https.service
cockpit-wsinstance-factory.socket
cockpit.socket
```

If we need http-redirect or http, we would get `cockpit-wsinstance@http.socket` (and similar), but only then.

martinpitt commented:
> Could we keep the fingerprint out of the cert-less instance?

That should be possible with some special-casing. It would add a tiny bit of extra C code. This is a possible "visual" simplification -- do you want to block on this, or should we look at that in a follow-up? (This whole story isn't complete yet anyway -- we at least need to bring back writing the certificates into /run, and solve a ws shutdown race condition that's already in master).

> Could we create the http and http-redirect socket units also on demand via the factory?

We need separate .service units for these as they call cockpit-ws with different arguments (`--no-tls` vs. `--proxy-tls-redirect` vs `--for-tls-proxy`). Technically we could look into encoding the entire command line into the instance name (after the @) and decoding it again in the unit, but that's also not exactly nice -- plus, that makes it much harder to see what these units do, or to override them.

mvollmer commented:
> do you want to block on this, or should we look at that in a follow-up?

Followup.

mvollmer commented:
> > Could we create the http and http-redirect socket units also on demand via the factory?
> 
> We need separate .service units for these as they call cockpit-ws with different arguments (`--no-tls` vs. `--proxy-tls-redirect` vs `--for-tls-proxy`).

Couldn't that be done by having a `cockpit-wsinstance@http.service` unit file?  Systemd would use that instead of using `cockpit-wsinstance@.service` to instantiate it, no?

mvollmer commented:
> Technically we could look into encoding the entire command line into the instance name (after the @) and decoding it again in the unit, but that's also not exactly nice -- plus, that makes it much harder to see what these units do, or to override them.

Or we add a `--instance=%i` option, and cockpit-ws main would look at that and recognize "http", "http-redirect", "https" and fingerprints, and set the right options.

allisonkarlitskaya commented:
> Could we keep the fingerprint out of the cert-less instance? Could we create the http and http-redirect socket units also on demand via the factory? Then the common case could look like

There's some history that's missing here:

Before this PR, we have three static instances (https, http, redirect).   I wanted to keep the no-cert https as a static instance and add dynamic instances only for the with-client-certificate case.  pitti and I went back and forth on this a few times and he managed to convince me, but I found it a bit weird to have one special case of a "https" or "nocert" or whatever name where all the other instances had the fingerprint as the instance name.  That's why we compromised on using a sha256 hash of the empty string to signify "no certificate".

If we're going to talk about turning *all* instances into dynamic instances of the same unit then it becomes more interesting to think about using non-sha256-blob names...  In that case, though, I guess i'd want to prefix the sha256 blobs in the client-cert case with some kind of a indicator that it's for client cert, so like:

 - wsinstance@http
 - wsinstance@nocert
 - wsinstance@clientcert-[sha256]
